### PR TITLE
WIP MS-1454: create save-and-exit behaviour to write to DB

### PR DESCRIPTION
**Pull request by jjrobothamho**
- create the behaviour
- create the service to manage data transit to and from the behaviour into a database
- create Postgres
- create Postgres database connection
- create Postgres data mapper to convert session data to a format db can digest
- create Postgres create function to write to db

### CODE REVIEW

File: storage/postgres/util/data-mapping.js [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    userEmail: data['user-email'],
+    jsonSavedData: data,
+  };
+
+  return mappedData;
+};
+
+module.exports = {
+  getData: getMappedData,
+};
```
Comment (jjrobothamho) [inline storage/postgres/util/data-mapping.js]: This file is going to be used convert our app data to a format for the database to injest @sulthan-ahmed

File: storage/postgres/postgres-db.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+ */
+async function create(data) {
+  try {
+    const mappedData = dataMapper.getData(data);
+    const { userEmail, jsonSavedData } = mappedData;
+
+    poolClient = await connect();
+
+    const resultId = poolClient.query(
+      queries.insertSessionData,
```
Comment (jjrobothamho) [inline storage/postgres/postgres-db.js]: Write function:

I have kept the query strings in another file @sulthan-ahmed

**Comment by jjrobothamho** [comment]
Plan to split this PR into 2 
- Postgres connection
- Save & return Behaviour

- this PR is more to see what it will look like combined
- plus to scrutinise the proposed structure 

@sulthan-ahmed

File: apps/nrm/services/save-and-exit-service.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+ *
+ * @returns {Promise} -  row id
+ */
+const writeToDB = async(data) => {
+  return await postgresDB.create(data);
+};
+
+module.exports = {
+  write: writeToDB,
+};
```
Comment (jjrobothamho) [inline apps/nrm/services/save-and-exit-service.js]: This file allows us to swap out different db's

File: storage/postgres/postgres-db.js [hunk 4]
```diff
@@ -0,0 +1,63 @@
+'use strict';
+
+const config = require('../../config');
+const pg = require('pg');
```
Comment (jjrobothamho) [inline storage/postgres/postgres-db.js]: pg is the default postgre package I've seen used
but need to research around to make sure it's the best choice and what implementation to use

https://node-postgres.com/api/pool

File: config.js [hunk 5]
```diff
@@ -18,6 +18,15 @@ module.exports = {
     port: process.env.REDIS_PORT || '6379',
     host: process.env.REDIS_HOST || '127.0.0.1'
   },
+  postgre: {
+    user: process.env.POSTGRE_USER || 'test',
+    host: process.env.POSTGRE_HOST || 'localhost',
+    password: process.env.POSTGRE_PASSWORD || 'test',
+    database: {
+      reports: process.env.POSTGRE_DB_NAME || 'postgres'
```
Comment (jjrobothamho) [inline config.js]: need to change this from postgres to reports or what every name you think the db should be called

File: apps/nrm/behaviours/save-and-exit.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+};
+
+module.exports = superclass => class extends superclass {
+  async saveValues(req, res) {
+    try {
+      const applicationId = await saveApplication(req);
+      // save resulting id to session
+      req.sessionModel.set('application-id', applicationId);
+
+      return res.redirect('/nrm/application-saved');
```
Comment (jjrobothamho) [inline apps/nrm/behaviours/save-and-exit.js]: These pages need to be added in a separate PR @sulthan-ahmed

File: package.json [hunk 7]
```diff
@@ -47,8 +47,10 @@
     "ms-uk-local-authorities": "^2.1.0",
     "ms-uk-police-forces": "^2.0.0",
     "ms-uk-regions": "^2.1.1",
+    "node-sass": "^4.13.0",
```
Comment (zuzak) [inline package.json]: Where's this dependency come from? Nothing else in this PR involves SASS.

Comment (jjrobothamho) [reply package.json]: May have to separate this out.
I have node 12 installed on this new machine so when I ran the app it required that I download the updated version.

Cheers

Comment (sulthan-ahmed) [reply package.json]: seems a bit weird to have to have this

**Comment by sulthan-ahmed** [comment]
@jjrobothamho do you want me to review these comments you made?

File: apps/nrm/behaviours/save-and-exit.js (hunk 6 above)
Comment (sulthan-ahmed) [reply apps/nrm/behaviours/save-and-exit.js]: I think it's fine so long as it doesn't fail.  You could just have a route for it in HOF even if it's a blank page

**Comment by sulthan-ahmed** [comment]
is this PR still relevant considering we are changing our approach?

File: storage/postgres/postgres-db.js (hunk 2 above)
Comment (sulthan-ahmed) [reply storage/postgres/postgres-db.js]: is this still relevant?

File: apps/nrm/behaviours/save-and-exit.js (hunk 6 above)
Comment (jjrobothamho) [reply apps/nrm/behaviours/save-and-exit.js]: cool cheers, I think I'll do that.
I'll ensure this code follow only the success flow when inserting data to db, if any errors then show alt page.

File: package.json (hunk 7 above)
Comment (jjrobothamho) [reply package.json]: Yeah I'll take this out. It updated on my end, no need unless our node version is going to be upgraded to 12. In that case it should be a separate PR.

**Comment by jjrobothamho** [comment]
Thanks guys I'll take heed of the comments and use them in the subsequent PR.
I'll close this for now.

File: package.json (hunk 7 above)
Comment (sulthan-ahmed) [reply package.json]: yeah and probably separate PR as well if that's the case

File: apps/nrm/behaviours/save-and-exit.js (hunk 6 above)
Comment (jjrobothamho) [reply apps/nrm/behaviours/save-and-exit.js]: @sulthan-ahmed for the routing how can we determine whether the save and exit button was clicked?

Comment (jjrobothamho) [reply apps/nrm/behaviours/save-and-exit.js]: maybe something in the params

Comment (sulthan-ahmed) [reply apps/nrm/behaviours/save-and-exit.js]: hmm interesting question, I'd probably make the save button a link and use url parameter to determine that they've clicked on the link.  I think that's why it's a link on other services with save and return.  I don't think HOF will capture it in the session

Comment (sulthan-ahmed) [reply apps/nrm/behaviours/save-and-exit.js]: I think that's what Joe Fitter did with this https://github.com/UKHomeOfficeForms/hof-behaviour-address-lookup
see the url here
https://www.ending-a-tenancy.homeoffice.gov.uk/property-address?step=manual

Comment (jjrobothamho) [reply apps/nrm/behaviours/save-and-exit.js]: Cheers, That's what I eventually implemented.
When they click on the link they are redirected and I then save the values in getValues() method.

I didn't set the params however. As I wanted to get the response back before determining the page content

Comment (sulthan-ahmed) [reply apps/nrm/behaviours/save-and-exit.js]: Ahh ok, that makes sense.  I guess you don't need to set the params then
